### ISSUE TITLE
JavaSymbolSolver to fail gracefully

### ISSUE THREAD

**Issue by tanin47** [javaparser/javaparser#2045]
Here's an example:

```
package org.codelab.sample;

public final class Test {
  public static int test2(Provider p) {
    return 3;
  }

  public void test() {
    test2(null);
  }
}
```

Please note that `Provider` isn't imported.

In the above case, when I visit `MethodCallExpr`, invoking `resolve()` will raise the exception `UnsolvedSymbolException` because it cannot resolve `Provider`.

I wonder if there's a way to not fail. Returning `JavaParserMethodDeclaration` would be helpful even though one of its parameters is invalid. It seems like the exception should have happened when I try to resolve that parameter instead.

Thank you.

**Comment by MysterAitch** [comment]
To resolve the method, the types of the parameters need to be resolved first (consider the case of overloaded methods e.g. `test(String p)` vs `test(Integer)`). 

Assuming it isn't present already, having the equivalent of `isParsed` (i.e. `isResolved`) and/or an `ResolveError` type instead of an inspection could be quite handy!

**Comment by tanin47** [comment]
I wasn't clear enough

> Returning JavaParserMethodDeclaration would be helpful even though one of its parameters is invalid

This would be helpful because it, at least, contains info like position, even though it can't resolve one of the parameters. To give you a context, I'm implementing the jump-to-definition feature.

> To resolve the method, the types of the parameters need to be resolved first (consider the case of overloaded methods e.g. test(String p) vs test(Integer)).

That makes sense. Some cases just can't be resolved. But something like `ResolveError` that contains some resolvable info (e.g. position, file,), when possible, would be hugely helpful.

**Comment by matozoid** [comment]
Makes sense to me, I've seen more complaints in this direction lately.

**Comment by ftomassetti** [comment]
Perhaps we could have a flag making the symbol solver more lenient and return an "unsolved method declaration". Would that work for you?

**Comment by tanin47** [comment]
> Perhaps we could have a flag making the symbol solver more lenient and return an "unsolved method declaration". Would that work for you?

Yes, unresolved method declarations that contains the position and the file (which can be obtained from the wrappedNode's root node) would help achieve my goal (which is a more torelant jump-to-definition feature).

I wonder if you can give me a pointer on how to make this improvement and the design that you have in mind. For example, would you return the position info through an exception or re-using the JavaParserMethodDeclaration or using a new class?

If I can understand it, I probably want to help. Thank you.

**Comment by tanin47** [comment]
I wonder if you can give me a pointer on how to make this improvement and the design that you have in mind. For example, would you return the position info through an exception or re-using the ResolvedMethodDeclaration or using a new class.

If I can understand it, I probably want to help. Thank you.
